File: Lab/Codes/MATLAB/MATLAB_spiking_neuron/neuron.m
classdef neuron %2
    properties %97
        C = 0.281; %4
        gL = 0.030; %5
        vR = -60.6; %6
        vT = -50.4; %7
        currentInj; %99
        T; %99
        timeseries; %99
        V; %99
    end %97
    methods %97
        function self = neuron %98
            self; %98
        end %98
        function self = create_injection_current(self) %8
            max_current=1;
            self.currentInj = cat(2,zeros(1,10),[[1:max_current*100]./100]); %9
            self.T = length(self.currentInj); %10
        end %97
        function self = leaky_integrate_and_fire(self) %11
            self.timeseries = linspace(1,self.T, self.T); %12
            self.V = ones(1,self.T).*self.vR %13
            ii = 1; %14
            while ii < (max(self.T)-2) %15
                dV = (-self.gL.*(self.V(ii) - self.vR) + ...
                    self.currentInj(ii))./self.C; %16
                self.V(ii+1) = self.V(ii)+dV; %17
                if self.V(ii+1) >= self.vT %18
                    self.V(ii+1) = 20; %19
                    self.V(ii+2) = self.vR; %20
                    ii = ii + 1; %21
                end %18
                ii = ii + 1; %22
            end
        end
        function plot_neuron(self) %23
            figure %24
            ax = subplot(2,1,1) ;%25
            plot(self.timeseries,self.currentInj,'color','k') %26
            title('Current Injection','FontAngle','italic') %27
            ylabel('Current (nA)') %28
            ax2 = subplot(2,1,2) ;%29

            plot(self.timeseries,self.V,'color','k') %30
            title('Integrate and Fire Voltage Response', 'FontAngle','italic') %31
            xlabel('Time (ms)') %32
            ylabel('Voltage (mV') %33
            %shg %35
            print('-dpng','-r300','Integrate and Fire Voltage Response'); %36
        end
    end
end

File: Lab/Codes/Python/Python_spiking_neuron/Neuron.py
import numpy as np
import matplotlib.pyplot as plt

class Neuron():
    def __init__(self):#intialise neuron variables
        self.C=0.281 #capacitance
        self.gL=0.030 #leak conductance
        self.vR=-60.6 #membrane potential
        self.vT=-50.4 #spike threshold

    def create_injection_current(self): #current injectipoon function
        max_current=1
        self.currentInj=np.append(np.zeros(10),np.arange(0,max_current,0.01))
        self.T=len(self.currentInj)

    def leaky_integrate_and_fire(self): #leaky-integrate-and-fire function
        self.timeseries = np.linspace(0,self.T-1,self.T)
        self.V=np.ones(self.T)*self.vR
        ii=0
        while ii<self.T-2:
            dV=(-self.gL*(self.V[ii]-self.vR)+self.currentInj[ii])/self.C
            self.V[ii+1]=self.V[ii]+dV
            if self.V[ii+1]>self.vT:
                self.V[ii+1]=20
                self.V[ii+2]=self.vR
                ii+=1
            ii+=1

    def plot_neuron(self): #plot neuron's membrane potential and current injection over time
        fig=plt.figure()
        fig.set_figwidth(10)
        fig.set_figheight(8)
        ax=fig.add_subplot(211)
        ax.plot(self.timeseries,self.currentInj,color='gold')
        ax.set_title('Current Injection',style='italic')
        ax.set_ylabel('Current (nA)', style='italic')

        ax2=fig.add_subplot(212)
        ax2.plot(self.timeseries,self.V,color='black')
        ax2.set_title('Integrate and Fire Membrane Potiental Response', style='italic')
        ax2.set_xlabel('Time (ms)',style='italic')        
        ax2.set_ylabel('Voltage (mV)', style='italic')
        plt.tight_layout()
        plt.rcParams.update({'font.size': 20})
        plt.show()
